hatchways/team-marinara
# gmail-auth api

Add the gmail auth-api back end which sends the user to Google to give Mail Sender read, compose and send access to their gmail account. Can be accessed at `/api/gmail-auth`

- Google API credential setup for Mail Sender can be seen and edited [here](https://console.developers.google.com/apis/credentials?project=mail-sender-1&folder=&organizationId=). This team should all have access.

- gmail-secret.json is required for this to run and is in .gitignore. It should be placed in the `/server` folder. 

- To test: 
  - Create gmail-secret.json in `/server` directory
  - Start server: `npm run dev`
  - Run `node /server/utils/populateUsers.js` to populate the Users collection
  - Visit http://localhost/api/gmail-auth in your browser
  - You should be taken through the authorization process and redirected to the home page
  - The first user in Users collection should now have a gmail_token field which can be used by readEmail() and sendEmail() functions to read and send emails from your account
  - Visiting the route again won't trigger authorization flow unless you delete the gmail_token field in Users collection

- Google OAuth required a public url to receive the request from and redirect to. Have used http://lvh.me, which is a public URL permanently pointing to 127.0.0.1 / localhost.

<img width="475" alt="Screenshot 2020-03-31 at 16 33 49" src="https://user-images.githubusercontent.com/42155812/78084577-6ae14880-736d-11ea-9be7-09028681645f.png">

- Because the app is not verified, a warning screen is displayed during authorization. User needs to click 'Advanced' and 'Go to lvh.me (unsafe)'
<img width="688" alt="Screenshot 2020-03-31 at 16 34 22" src="https://user-images.githubusercontent.com/42155812/78084602-7df41880-736d-11ea-9521-137ee5f0848d.png">

<img width="467" alt="Screenshot 2020-03-31 at 16 35 38" src="https://user-images.githubusercontent.com/42155812/78084667-ab40c680-736d-11ea-8a7d-fb8e3c3f4c3b.png">

- userId (_id from users collection) is currently hardcoded awaiting completion of login/registration and autorization back-end. Expecting userId to be provided in req.header payload.

- Added 2 test functions to send and read user's email using supplied tokens. These are not used now, but can be used when we implement those functions.

- Tidy up routes: moved all routes to routes/index.js
